src/corehost/cli/fxr/hostfxr.cpp

```diff
@@ -274,6 +276,167 @@ SHARED_API int32_t hostfxr_resolve_sdk(
     return cli_sdk.size() + 1;
 }
 
+typedef void (*hostfxr_resolve_sdk2_result_fn)(
+    const pal::char_t* resolved_sdk_dir, 
+    const pal::char_t* global_json_path);
+
+enum hostfxr_resolve_sdk2_flags_t : int32_t
+{
+    disallow_prerelease = 0x1,
+};
+
+//
+// Determines the directory location of the SDK accounting for
+// global.json and multi-level lookup policy.
+//
+// Invoked via MSBuild SDK resolver to locate SDK props and targets
+// from an msbuild other than the one bundled by the CLI.
+//
+// Parameters:
+//    exe_dir
+//      The main directory where SDKs are located in sdk\[version]
+//      sub-folders. Pass the directory of a dotnet executable to
+//      mimic how that executable would search in its own directory.
+//      It is also valid to pass nullptr or empty, in which case
+//      multi-level lookup can still search other locations if 
+//      it has not been disabled by the user's environment.
+//
+//    working_dir
+//      The directory where the search for global.json (which can
+//      control the resolved SDK version) starts and proceeds
+//      upwards. 
+//
+//   flags
+//      Bitwise flags that influence resolution.
+//         disallow_prerelease (0x1)
+//           do not allow resolution to return a prerelease SDK version 
+//           unless  prerelease version was specified via global.json.
+//
+//   result
+//      Callback invoked to return the resolved sdk directory and 
+//      global.json path used.
+//      
+//      If resolution succeeds, resolved_sdk_dir will hold the
+//      path to the resolved SDK director, otherwise it will
+//      be null.
+//
+//      If global.json is used, then global_json_path will hold
+//      the path to global.json. If there was global.json found,
+//      or the contents of global.json did not impact resolution
+//      (e.g. no version specified), then global_json_path will
+//      be null.
+//
+// Return value:
+//   0 on success, otherwise failure
+//   0x8000809b - SDK could not be resolved (SdkResolverResolveFailure)
+// 
+// String encoding:
+//   Windows     - UTF-16 (pal::char_t is 2 byte wchar_t)
+//   Unix        - UTF-8  (pal::char_t is 1 byte char)
+//
+SHARED_API int32_t hostfxr_resolve_sdk2(
```

Comment:
> Given the material difference between hostfxr_resolve_sdk and this method is that the location of global.json is returned

The flags (disallow_prerelease) are also material.

> did we consider adding a 'return the path to the global.json' instead of adding a new overload?

I have considered this, but my concerns with that are:

1. Perf: Looking for global.json requires probing on disk from working directory up to root, and I don't want to do this once for the resolution and again for the logging. I would also like to be able to log the path on success paths too because it is common that users do not realize that they have a leftover global.json and that has made.

2. There's a possible race between "resolve the SDK", and "where is global.json" such that the global.json returned by the second query is not the same as the one that was actually used during resolution.

Reply:
I would prefer a place where we land either with smaller building blocks (perf and a potential race could be interesting, though do we have any evidence that is a concern?) or a new API that is more expandable.  If we stay with an SDK2 version, I think we should do a bit more to help it be more expandable in the future.  This shared_api is really just a thin wrapper over the resolve_sdk_dotnet_path.  I like the addition of flags.  Can we come up with a strategy that enables us to return more than just the two paths?  Perhaps either a property bag, or have the function callback take a key and a value and then call it multiple times.  I don't think we need to do much future proofing, but a little more to help us as things may evolve in intuitive ways.

Reply:
> Perhaps either a property bag, or have the function callback take a key and a value and then call it multiple times. I don't think we need to do much future proofing, but a little more to help us as things may evolve in intuitive ways.

Very good call.

Reply:
I think I'll go with the multiple callbacks because it's more flexible. We could even wire tracing to it at some point, which is something I've wanted. Right now, you can set COREHOST_TRACE and get the logging to stderr, but it would be nice if you could set a tracing flag on the API and get the verbose diagnostics sent to the msbuild log.